core/trino-main/src/main/java/io/trino/server/testing/exchange/LocalFileSystemExchangeSource.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.server.testing.exchange;
+
+import com.google.common.collect.ImmutableList;
+import io.airlift.slice.InputStreamSliceInput;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceInput;
+import io.airlift.units.DataSize;
+import io.trino.spi.exchange.ExchangeSource;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.UncheckedIOException;
+import java.nio.file.Path;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import static io.airlift.units.DataSize.Unit.KILOBYTE;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+public class LocalFileSystemExchangeSource
+        implements ExchangeSource
+{
+    private static final int BUFFER_SIZE_IN_BYTES = toIntExact(DataSize.of(4, KILOBYTE).toBytes());
+
+    @GuardedBy("this")
```

Comment:
Final variables don't need `@GuardedBy` annotation

Reply:
`Iterator` is mutable and has to be accessed under a lock